### PR TITLE
HRIN-672: Added ticket synchonization

### PR DESCRIPTION
**Pull request by rimi-itk**
https://jira.itkdev.dk/browse/HRIN-672

Adds endpoint for synchronizing just a single ticket (rather that all tickets on the hearing).

### CODE REVIEW

File: web/modules/custom/hoeringsportal_deskpro/src/Service/HearingHelper.php [hunk 1]
```diff
@@ -414,6 +382,142 @@ public function synchronizeHearingTickets(Node $hearing) {
     return ['hearing' => $hearing->id(), 'data' => $data];
   }
 
+  /**
+   * Get data from Deskpro and store in hearing node.
+   */
+  public function synchronizeTicket(array $payload, $delayed = TRUE) {
+    [$hearing, $ticketId] = $this->validateTicketPayload($payload);
```
Comment (cableman) [inline web/modules/custom/hoeringsportal_deskpro/src/Service/HearingHelper.php:389]: `validateTicketPayload` may throw `RuntimeException` are these handled?

**Review by cableman — CHANGES_REQUESTED** [review]

File: web/modules/custom/hoeringsportal_deskpro/src/Service/HearingHelper.php (hunk 1 above)
Comment (rimi-itk) [reply web/modules/custom/hoeringsportal_deskpro/src/Service/HearingHelper.php:389]: Not really, but we don't really care about handling invalid payloads in a nice way, i.e. we expect them to be valid.

File: web/modules/custom/hoeringsportal_deskpro/src/Service/HearingHelper.php [hunk 2]
```diff
@@ -414,6 +382,142 @@ public function synchronizeHearingTickets(Node $hearing) {
...
+      $queue = Queue::load('hoeringsportal_deskpro');
+      $delay = max($this->deskpro->getConfig()->getSynchronizationDelay(), 60);
+
+      // Remove any already queued jobs for same hearing and ticket from the
+      // queue.
+      $backend = $queue->getBackend();
+      $hearingIdFieldName = $this->getHearingIdFieldName();
+      $matchesCurrentPayload = function (array $otherPayload) use ($payload, $hearingIdFieldName) {
+        // Payloads must match on both ticket.id and ticket.$hearingIdfieldName.
+        return isset($payload['ticket']['id'], $otherPayload['ticket']['id'])
```
[the remark this answers is not in the document]
Comment (rimi-itk) [reply web/modules/custom/hoeringsportal_deskpro/src/Service/HearingHelper.php:400]: No. I'll clean up.

**Review by cableman — APPROVED** [review]

File: docker-compose.server.yml [hunk 3]
```diff
@@ -16,7 +16,7 @@ services:
       - app
     environment:
       - PHP_MAX_EXECUTION_TIME=30
-      - PHP_MEMORY_LIMIT=128M
+      - PHP_MEMORY_LIMIT=512M
```
Comment (cableman) [inline docker-compose.server.yml:19]: What is the reason behind this increase in memory requirements. Even DDB CMS can run under 256 MB memory. If this is really an requirement I think we should analyse the code and figure out why.

File: web/modules/custom/hoeringsportal_deskpro/src/Service/HearingHelper.php [hunk 4]
```diff
@@ -447,20 +447,44 @@ public function runSynchronizeTicket(array $payload) {
...
-
-    if (NULL === $ticketIndex) {
-      $tickets[] = $ticket;
+    // Remove unpublished or deleted tickets.
+    $remove =
+      // Ticket is unpublished if and only if the field "unpublish_reply" === 1.
+      1 === (int) ($ticket['fields']['unpublish_reply'] ?? NULL)
+      // A deleted ticket has a status starting with "hidden" (and is not really
+      // deleted so it makes sense to get it from the api).
+      || 0 === strpos($ticket['status'], 'hidden');
```
Comment (cableman) [inline web/modules/custom/hoeringsportal_deskpro/src/Service/HearingHelper.php:456]: Are we still in PHP 7.x or `str_starts_with` could be used.

**Review by cableman — CHANGES_REQUESTED** [review]

File: web/modules/custom/hoeringsportal_deskpro/src/Service/HearingHelper.php (hunk 4 above)
Comment (rimi-itk) [reply web/modules/custom/hoeringsportal_deskpro/src/Service/HearingHelper.php:456]: We're still on php 7.4 (but hopefully not for long).

**Pull request merged**